src/Moq/SetupCollection.cs

```diff
@@ -11,17 +11,35 @@ internal sealed class SetupCollection
 	{
 		private List<Setup> setups;
 		private uint overridden;  // bit mask for the first 32 setups flagging those known to be overridden
+		private bool hasEventSetup;
 
 		public SetupCollection()
 		{
 			this.setups = new List<Setup>();
 			this.overridden = 0U;
+			this.hasEventSetup = false;
+		}
+
+		public bool HasEventSetup
+		{
+			get
+			{
+				lock(this.setups)
+				{
+					return this.hasEventSetup;
```

Comment:
I agree with you, this, however, requires some changes in the `RemoveAll` to avoid multiple writes to `hasEventSetup`(currently I set it to `false` at the beginning and then, potentially, update in each iteration, I think update should be done only once as a very last step in the method).